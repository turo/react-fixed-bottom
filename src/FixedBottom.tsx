import React, {ReactElement, RefObject} from 'react';
import throttle from 'lodash.throttle';
import {isSafariMobile} from './util';
import {Cancelable} from 'lodash';

const SAFARI_MOBILE_BOTTOM_MENU_HEIGHT = 44;

export interface FixedBottomProps {
  children: React.ReactChild;
  offset?: number;
}

export interface FixedBottomState {
  bottom: number;
}

/**
 * Component that allows sticking elements to the bottom of the screen:
 *   - iOS Safari has a long-standing issue when using `position: fixed` near the
 *   bottom edge of the screen.
 *   - `window.innerHeight` reflects the available size and it changes when the
 *   navigation bars are visible.
 *
 * This components implements the following solution:
 *   1. Check for `iosSafariMobile()`
 *   2. Check if the element is hidden by the overflow bar
 *   3. Either add a `44px` offset to place it above the bar or
 *   keep it at its original position
 *   4. On mount, check whether the overflow bar is already
 *   visible
 */
export class FixedBottom extends React.PureComponent<FixedBottomProps> {

  state: FixedBottomState = {
    bottom: this.props.offset || 0,
  };

  isSafariMobile = isSafariMobile();

  anchorRef: RefObject<HTMLDivElement> = React.createRef();

  deferredComputeOffsetBottom: number;

  handleScroll: (() => void) & Cancelable;

  constructor(props: FixedBottomProps) {
    super(props);
    this.handleScroll = throttle(this.computeOffsetBottom, 200);
  }

  componentDidMount() {
    if (this.isSafariMobile) {
      window.addEventListener('scroll', this.handleScroll);
      this.deferredComputeOffsetBottom = setTimeout(this.computeOffsetBottom);
    }
  }

  componentWillUnmount() {
    if (this.isSafariMobile) {
      this.handleScroll.cancel();
      window.removeEventListener('scroll', this.handleScroll);
      window.clearTimeout(this.deferredComputeOffsetBottom);
    }
  }

  computeOffsetBottom = () => {
    if (!this.anchorRef.current) {
      return;
    }

    const {bottom} = this.anchorRef.current.getBoundingClientRect();
    const {offset = 0} = this.props;
    if (Math.floor(Number(bottom)) > window.innerHeight) {
      this.setState({bottom: offset + SAFARI_MOBILE_BOTTOM_MENU_HEIGHT});
    } else {
      this.setState({bottom: offset});
    }
  };

  render() {
    const {bottom} = this.state;
    const {children, offset} = this.props;
    const node = React.cloneElement(children as ReactElement, {
      style: {
        ...(children as ReactElement).props.style,
        bottom,
        position: 'fixed',
      },
    });
    return (
      <>
        {node}
        {this.isSafariMobile && (
          <div
            ref={this.anchorRef}
            style={{
              position: 'fixed',
              bottom: offset,
            }}
          />
        )}
      </>
    );
  }
}
