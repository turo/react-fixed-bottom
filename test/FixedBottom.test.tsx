import 'jest';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';

import {configure, shallow, ShallowWrapper} from 'enzyme';
import {FixedBottom, FixedBottomProps, FixedBottomState} from '~/FixedBottom';

configure({adapter: new Adapter()});

const offset = 20;
const childElementProps = {
  children: 'Dummy content',
  className: 'childElement',
  style: {top: 20},
};
const childElement = <div {...childElementProps} />;

describe('<FixedBottom /> tests', () => {
  let wrapper: ShallowWrapper<any, FixedBottomState, FixedBottom>;
  let instance: FixedBottom;

  beforeEach(() => {
    wrapper = shallow(<FixedBottom offset={offset}>{childElement}</FixedBottom>);
    instance = wrapper.instance();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('it should render the children with expected props', () => {
    const component = renderer.create(<FixedBottom offset={offset}>{childElement}</FixedBottom>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('handleScroll should throttle computeOffsetBottom', () => {
    jest.mock('lodash.throttle');
    const mockedThrottle = require('lodash.throttle');
    mockedThrottle.mockImplementation(fn => fn);
    wrapper = shallow(<FixedBottom offset={offset}>{childElement}</FixedBottom>);
    instance = wrapper.instance();
    const mockDiv = document.createElement('div');
    mockDiv.getBoundingClientRect = jest.fn().mockReturnValue({bottom: window.innerHeight});
    instance.anchorRef = {
      current: mockDiv,
    };
    instance.handleScroll();
    expect(mockDiv.getBoundingClientRect).toHaveBeenCalledTimes(1);
  });

  describe('componentDidMount', () => {
    let addEventListenerSpy: jest.SpyInstance<void, any>;
    let computeOffsetBottomStub: jest.Mock<any, any>;

    beforeEach(() => {
      jest.useFakeTimers();
      addEventListenerSpy = jest.spyOn(window, 'addEventListener');
      computeOffsetBottomStub = jest.fn();
      instance.computeOffsetBottom = computeOffsetBottomStub;
    });

    test('should bind a scroll listener, and defer compute offsetBottom if it is safari mobile', () => {
      instance.isSafariMobile = true;
      instance.componentDidMount();
      expect(addEventListenerSpy).toBeCalledWith('scroll', instance.handleScroll);
      expect(computeOffsetBottomStub).not.toBeCalled();
      jest.advanceTimersByTime(1);
      expect(computeOffsetBottomStub).toBeCalled();
    });

    test('should do nothing otherwise', () => {
      instance.isSafariMobile = false;
      instance.componentDidMount();
      expect(addEventListenerSpy).not.toBeCalled();
      expect(computeOffsetBottomStub).not.toBeCalled();
    });
  });

  describe('componentWillUnmount', () => {
    let clearTimeoutSpy: jest.SpyInstance<void, [(number | undefined)?]>;
    let removeEventListenerSpy: jest.SpyInstance<void, [string, EventListenerOrEventListenerObject, (boolean | EventListenerOptions | undefined)?]>;

    beforeEach(() => {
      removeEventListenerSpy = jest.spyOn(window, 'removeEventListener');
      jest.spyOn(instance.handleScroll, 'cancel');
      clearTimeoutSpy = jest.spyOn(window, 'clearTimeout');
    });

    test('should remove the scroll listener and cancel the scroll handler if it is safari mobile', () => {
      instance.isSafariMobile = true;
      instance.componentWillUnmount();
      expect(removeEventListenerSpy).toBeCalledWith('scroll', instance.handleScroll);
      expect(instance.handleScroll.cancel).toHaveBeenCalledTimes(1);
      expect(clearTimeoutSpy).toHaveBeenCalledTimes(1);
    });

    test('should do nothing otherwise', () => {
      instance.isSafariMobile = false;
      instance.componentWillUnmount();
      expect(removeEventListenerSpy).not.toBeCalled();
      expect(instance.handleScroll.cancel).not.toBeCalled();
      expect(clearTimeoutSpy).not.toBeCalled();
    });
  });

  describe('computeOffsetBottom', () => {
    test('should not fail if there is no ref.current', () => {
      const computingWithoutRefCurrent = () => {
        instance.anchorRef = {current: null};
        instance.computeOffsetBottom();
      };
      expect(computingWithoutRefCurrent).not.toThrow();
    });

    test('should add the safari menu height to the offset if it is hiding the element', () => {
      const mockDiv = document.createElement('div');
      mockDiv.getBoundingClientRect = jest.fn().mockReturnValue({bottom: window.innerHeight + 10});
      instance.anchorRef = {
        current: mockDiv
      };
      expect(wrapper.state().bottom).toBe(offset);
      instance.computeOffsetBottom();
      expect(wrapper.state().bottom).toBe(offset + 44);
    });

    test('should use the provided offset otherwise', () => {
      const mockDiv = document.createElement('div');
      mockDiv.getBoundingClientRect = jest.fn().mockReturnValue({bottom: window.innerHeight - offset});
      instance.anchorRef = {
        current: mockDiv,
      };
      expect(wrapper.state().bottom).toBe(offset);
      instance.computeOffsetBottom();
      expect(wrapper.state().bottom).toBe(offset);
    });
  });

  describe('render', () => {
    test('should clone the children and add some extra props', () => {
      const clone = wrapper.find(`.${childElementProps.className}`);
      expect(clone.type()).toBe(childElement.type);
      expect(clone.props().style).toEqual({
        ...childElementProps.style,
        bottom: offset,
        position: 'fixed',
      });
    });

    test('should render an anchor div stuck to bottom when it is safari mobile', () => {
      expect(wrapper.find('div').length).toBe(1);
      instance.isSafariMobile = true;
      wrapper.setProps({reRender: 1});
      expect(wrapper.find('div').length).toBe(2);
      expect(
        wrapper
          .find('div')
          .at(1)
          .props().style
      ).toEqual({
        position: 'fixed',
        bottom: offset,
      });
    });

    test('should have an offset of 0 by default', () => {
      const defaultPropsWrapper: ShallowWrapper<FixedBottomProps, FixedBottomState, FixedBottom> = shallow(
        <FixedBottom>
          <div>hello</div>
        </FixedBottom>
      );
      expect(defaultPropsWrapper.state().bottom).toBe(0);
    });
  });
});
