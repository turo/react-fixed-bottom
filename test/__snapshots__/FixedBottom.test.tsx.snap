// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FixedBottom /> tests it should render the children with expected props 1`] = `
<div
  className="childElement"
  style={
    Object {
      "bottom": 20,
      "position": "fixed",
      "top": 20,
    }
  }
>
  Dummy content
</div>
`;
